configure: Enable tcmalloc to use a tmcalloc library

mohit84 opened the pull request:
The data is already available on github issue(#2771)
specific to getting improvement after enable tcmalloc

Fixes: #2771
Change-Id: I29256db870185d521d9b9adc2b7fd66646ed1c56
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mykaul commented on glusterfs.spec.in:
```diff
@@ -260,12 +266,14 @@ BuildRequires:    libasan
 BuildRequires:    libxml2-devel openssl-devel
 BuildRequires:    libaio-devel libacl-devel
 BuildRequires:    python%{_pythonver}-devel
+BuildRequires:    gperftools-devel
```
@msaju - that will require changes in Gluster-builds Dockerfiles as well.

msaju replied:
@mykaul  - noted, will do the needful

dmantipov commented:
Shouldn't `BuildRequires:    gperftools-devel` be under `%if` in case of someone wants to build the package `--without tcmalloc`?

mohit84 commented:
> Shouldn't `BuildRequires: gperftools-devel` be under `%if` in case of someone wants to build the package `--without tcmalloc`?

I have not provided without-tcmalloc option because a user will face regression compare to previous release.If we want to provide this option we have to revert  the patch #2675 , for more please refer #2771

dmantipov commented:
> I have not provided without-tcmalloc option because a user will face regression compare to previous release.If we want to provide this option we have to revert the patch #2675 , for more please refer #2771

Hmm... `smallfile` (and everything that works via `FUSE`) isn't the only use case. Did you try native `gfapi` clients like `fio`, `qemu` etc.? Anyway I would prefer more flexibility for package builders who really knows what they're doing.

mohit84 commented:
> > I have not provided without-tcmalloc option because a user will face regression compare to previous release.If we want to provide this option we have to revert the patch #2675 , for more please refer #2771
> 
> Hmm... `smallfile` (and everything that works via `FUSE`) isn't the only use case. Did you try native `gfapi` clients like `fio`, `qemu` etc.? Anyway I would prefer more flexibility for package builders who really knows what they're doing.

Why do u believe the behavior will be different for gfapi or qemu, afaik at fuse xlator level there is no much mem_get call, the other common xlators like afr,dht,client are having huge impact due to mempool.

dmantipov commented:
> Why do u believe the behavior will be different for gfapi or qemu, afaik at fuse xlator level there is no much mem_get call, the other common xlators like afr,dht,client are having huge impact due to mempool.

This is a question of an extensive benchmarking. In general, my points around the whole topic are:

1. Memory pools was an overengineered layer between system `malloc()` and upper layers, so #2675 should not be reverted unless something extraordinary happened.
2. I'm not against `tcmalloc` by default since it tends to be faster, scales better (maybe at the cost of consuming more memory) etc., etc., etc.
3. Nevertheless `--without tcmalloc` package build option costs almost nothing and should be provided.

mohit84 commented:
> > Why do u believe the behavior will be different for gfapi or qemu, afaik at fuse xlator level there is no much mem_get call, the other common xlators like afr,dht,client are having huge impact due to mempool.
> 
> This is a question of an extensive benchmarking. In general, my points around the whole topic are:
> 
> 1. Memory pools was an overengineered layer between system `malloc()` and upper layers, so [libglusterfs: drop core infrastructure of memory pools #2675](https://github.com/gluster/glusterfs/pull/2675) should not be reverted unless something extraordinary happened.
> 2. I'm not against `tcmalloc` by default since it tends to be faster, scales better (maybe at the cost of consuming more memory) etc., etc., etc.
> 3. Nevertheless `--without tcmalloc` package build option costs almost nothing and should be provided.

I am not against to without-libtcmalloc option, What i mean to say if we do provide without-libtcmalloc
option and a user build the rpms on release-10 and when a user will compare the numbers from release-(7/8/9)
the user will face regression.To avoid a regression there are two ways 
Either 1) Don't provide without-tcmalloc option
   or
 2) Provide a without-tcmalloc option after revert #2675 sothat if a user wants to use glibc based pool it is better to use glusterfs based mempool.

@amarts / @xhernandez 
  Can you please share the views on the same?

amarts commented:
My thoughts are like this:

* We don't need memory-pool in the long run, so #2675 stays (ie, no revert).
* considering tcmalloc is significantly better performing (in some cases >35%), as a project using that by default is a good idea.
* The way I understand @dmantipov's comment is, we need to keep tcmalloc optional (which is done in the PR), and extend that 'optional' feature to building rpm. rpms build by default parameters will have tcmalloc, and if someone wants to build without it, they should be provided an option - which IMO is good thing to have. (ie, any option in configure should be extended to rpmbuild).

mohit84 commented:
> My thoughts are like this:
> 
> * We don't need memory-pool in the long run, so [libglusterfs: drop core infrastructure of memory pools #2675](https://github.com/gluster/glusterfs/pull/2675) stays (ie, no revert).
> * considering tcmalloc is significantly better performing (in some cases >35%), as a project using that by default is a good idea.
> * The way I understand @dmantipov's comment is, we need to keep tcmalloc optional (which is done in the PR), and extend that 'optional' feature to building rpm. rpms build by default parameters will have tcmalloc, and if someone wants to build without it, they should be provided an option - which IMO is good thing to have. (ie, any option in configure should be extended to rpmbuild).

Thanks amar for sharing your view, I will wait for xavi response also if he also believes the same i will provide an option to build rpms without tcmalloc.

msaju commented:
Agree with @amarts except for comment 3.
Providing a without-tcmalloc option; we will need to revert #2675, else its a regression.

amarts commented:
Not able to understand the reasoning.

Technically, there are 2 things here (technically separate).

1. Drop memory pool implementation altogether - Goes in as a bullet point in release notes.
2. Make tcmalloc as default - goes in as a bullet point in release.

Now, for 2, there would be an option for 'some' users who don't want tcmalloc. They technically would be using rpms of their own, as community rpms would be having tcmalloc. Now, for them there would be regression of performance. Which is OK. They are using it against recommended practice, and hence the regression is justified.

amarts commented:
also #2696 is a possible PR to consider, where we can add different malloc implementation for users to consider. By default would be tcmalloc, but users can pick any of their choice.

msaju commented:
> Not able to understand the reasoning.
> 
> Technically, there are 2 things here (technically separate).
> 
> 1. Drop memory pool implementation altogether - Goes in as a bullet point in release notes.
> 2. Make tcmalloc as default - goes in as a bullet point in release.
> 
> Now, for 2, there would be an option for 'some' users who don't want tcmalloc. They technically would be using rpms of their own, as community rpms would be having tcmalloc. Now, for them there would be regression of performance. Which is OK. They are using it against recommended practice, and hence the regression is justified.

From a product perspective, its better not to recommend something which regress the performance; especially when we have alternatives. And if we really need to;  then lets revert the patch and add the feature.

amarts commented:
> From a product perspective, its better not to recommend something which regress the performance; especially when we have alternatives. And if we really need to; then lets revert the patch and add the feature.

I am loss to understand where we are 'recommending' anything which regress performance.

We have 400+ options in glusterfs, but each of those have a default value, which we (developers) think are good, or sane. Doesn't mean users can set it to anything they want. We recommend only values which are well suited. Even in this case, we are 'recommending' and keeping tcmalloc as default. No other options are recommended.

An option is given in a project so the things we haven't tested, or figured out can be tested by others easily without making code changes.

In this case, we are debating a simple thing. An option provided in `./configure` to be available as an option in `rpmbuild`. Performance regression or other things are not at all a discussion point for that particular ask, IMO. 

For example, tcmalloc was an option available for last 2 yrs. But we are making it default now. Similarly someone may come up with something else in future if things are optional.

xhernandez commented:
I think that making the dependency optional when we know that without tcmalloc we have significant regressions for almost all operations doesn't make much sense because no one should use it. However, if that's the case, why we do even provide that option ?

If it's mandatory, we should remove the configure option and always check for the presence of the library to compile the code. If it's not mandatory and the user disables it, I think it doesn't make sense to add a dependency to a library that the code won't use.

We are discussing two different things here:

1. Should we make the requirement optional ?

I think yes. If we provide an option for tcmalloc and it can be disabled, the dependency won't be needed. Otherwise we shouldn't provide the option in the first place and make compilation against tcmalloc a requirement. In this case the dependency will always be needed unconditionally.

2. Should we allow to compile GlusterFS using the glibc allocator ?

That's a more complex question.

we need to consider that we cannot test all possible workloads, environments and platforms, so it might happen that for some specific cases this regression doesn't happen (or it's not so big) using the glibc allocator (though I would be very surprised). Also, tcmalloc is not the only alternative for a memory allocator. There are others that could theoretically perform even better than tcmalloc. For this reason, #2696 could make sense.

#2696 provides a generic solution for selecting any allocator. There we could decide which allocators we allow and which ones we don't allow, and it can be updated over time if things change, but in all cases the dependencies to the related libraries should be added only when needed (i.e. enabled by the option).

dmantipov commented:
Again, the whole thing is simple: let's make `tcmalloc` the default but take #2696 and corresponding `--with / --without` RPM spec options for the brave builders who understands what they're doing.

It's also desirable for configure to emit a warning if an allocator other than `tcmalloc` is enabled.

mohit84 commented:
/run regression

mohit84 commented:
/run regression

dmantipov commented:
Note #2696 is updated to choose tcmalloc by default but add '--with jemalloc' and '--with system-malloc' RPM spec options.

xhernandez commented:
I've discussed this with @msaju and @mohit84 in more detail and I've realized that I've missed an important point...

The problem right now is that if the user chooses to use system malloc, it will suffer a significant regression. What I missed is that we can prevent this regression while still allowing the user to choose the system allocator.

The idea is to revert #2675 and write a new patch that makes the use of mempools optional instead of completely dropping them (@mohit84 can work on that). Now we can let the user choose:

* If tcmalloc is choosen, we also disable mempools.
* If system malloc is choosen, we enable mempools.

This way we make sure that there won't be regressions in any case.

We also need to decide if the user will be able to explicitly enable/disable the mempools. Given that by default we will use tcmalloc, the only possibility for a regression would be to explicitly choose the system allocator *and* disable mempools. I think whoever does that needs to know what he's doing (standard packages from distributions will use default settings), and it could be useful for debugging purposes in the future.

@amarts, @dmantipov are you ok with making mempools optional instead of completely removing them ? if things work well and we acquire some experience without the mempools, we could consider to remove them in a future version of Gluster.

dmantipov commented:
> The problem right now is that if the user chooses to use system malloc, it will suffer a significant regression.

Has somebody really tried to tune system (assuming glibc, e.g. `ptmalloc`) for our purposes? Has somebody really tried this, for example:
```
When employing newer glibc versions, applications may in some cases exhibit high contention when accessing 
arenas.  In these cases, it may  be  beneficial  to  increase  M_ARENA_MAX  to  match the number of threads.
This is similar in behavior to strategies taken by tcmalloc and jemalloc (e.g., per-thread allocation pools).
```
Relying on _this malloc is too slow, let's use another one_ is not an option for the serious benchmark. You should know your `malloc` and must ensure that you've taken everything from it.

xhernandez commented:
> > The problem right now is that if the user chooses to use system malloc, it will suffer a significant regression.
> 
> Has somebody really tried to tune system (assuming glibc, e.g. `ptmalloc`) for our purposes? Has somebody really tried this, for example:
> 
> ```
> When employing newer glibc versions, applications may in some cases exhibit high contention when accessing 
> arenas.  In these cases, it may  be  beneficial  to  increase  M_ARENA_MAX  to  match the number of threads.
> This is similar in behavior to strategies taken by tcmalloc and jemalloc (e.g., per-thread allocation pools).
> ```

I'm not aware of any such test, but I agree we could investigate this further. However right now Gluster with glibc allocator runs slower without mempools. That's a fact. That's why making mempools optional is proposed. If/when we optimize Gluster for glibc, if we see that mempools are not really needed (i.e. there's no regression with glibc malloc and without mempools), then we can completely wipe them from the code. Until then I think it makes sense to keep the mempools as an option.

dmantipov commented:
Well, I'm not strictly against the reverting #2675. But, purely from the technical point of view, it makes sense only if our **current** best-bet allocator, `tcmalloc`, doesn't work well enough on some platform we're interesting in (old Linux, *BSD, whatever). Otherwise it's enough to enforce `tcmalloc` in `configure`, RPM spec, etc. wherever required.

amarts commented:
> @amarts, @dmantipov are you ok with making mempools optional instead of completely removing them ?

I am OK. But I consider that as an extra effort which is not worth for solving something which is not going to be used. But if people are planning to work on it, fine.

> The problem right now is that if the user chooses to use system malloc, it will suffer a significant regression

Exactly my point. How many users are there ?  How many of the ./configure options are used by anyone outside of the recommended way so far? 

All I am saying is as we have a combination of better performing results, why bother on that 1 (or say 5) users who would need it. They can always revert the patch manually if required.

Anyways, my suggestion was based considering the time all of this needs, and considering we have fewer people on the project, they can pick up other critical tasks. 

---
If you ask me honestly, if I had all the resources at my expense, yes, I would revert the commit removing mempool as it introduces regression, and work as you suggested.

xhernandez commented:
Then both of you are suggesting to make tcmalloc mandatory. Am I right ?

In this case #2696 is not needed and this patch should make tcmalloc non-optional, and fail at configure time if it's not present, right ?

For me that's also ok if you prefer that.

amarts commented:
* to mandate tcmalloc, we have some infra changes to take care of. 
* i personally like the support to have different malloc libraries as option in #2696 

How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.

mohit84 commented:
> * to mandate tcmalloc, we have some infra changes to take care of.
> * i personally like the support to have different malloc libraries as option in [build, libglusterfs: add --enable-malloc option #2696](https://github.com/gluster/glusterfs/pull/2696)
> 
> How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.

We can make it default choice only for linux environment.

xhernandez commented:
> * to mandate tcmalloc, we have some infra changes to take care of.
> * i personally like the support to have different malloc libraries as option in [build, libglusterfs: add --enable-malloc option #2696](https://github.com/gluster/glusterfs/pull/2696)
> 
> How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.

If we do this, your first point is still a problem (the infra needs to be updated for sure).

For the second point, I would also like to provide support for different malloc libraries specially for debugging and experimenting but, as you said, the number of people that could make use of that (in a useful way) is quite limited, and they are perfectly able to do those changes by themselves (as reverting the mempools patch if they want to). So, why providing an option that we know it will regress if we don't keep mempools around ?

@msaju @mohit84 what do you think ?

mykaul commented:
General remarks:
- merging the patch to drop memory pools, leaves us with devel branch running ~8.5% lower performance on average (in our standard tests). This is not the ideal situation. Ideally, we enable performance enhancements first (such as enabling tcmalloc by default), then drop memory pools, so we are not causing a regression.
- I wasn't sure we've reached a consensus we should drop them altogether. I'm personally in favor (though I've found some minor improvements that could be made), but it wasn't clear it was widely agreed upon.

mohit84 commented:
> > * to mandate tcmalloc, we have some infra changes to take care of.
> > * i personally like the support to have different malloc libraries as option in [build, libglusterfs: add --enable-malloc option #2696](https://github.com/gluster/glusterfs/pull/2696)
> > 
> > How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.
> 
> If we do this, your first point is still a problem (the infra needs to be updated for sure).
> 
> For the second point, I would also like to provide support for different malloc libraries specially for debugging and experimenting but, as you said, the number of people that could make use of that (in a useful way) is quite limited, and they are perfectly able to do those changes by themselves (as reverting the mempools patch if they want to). So, why providing an option that we know it will regress if we don't keep mempools around ?
> 
> @msaju @mohit84 what do you think ?

I agree with you, we should provide an system_malloc option only while we have mempools. All of us wants to support for different malloc libraries the only thing we need to make mempool optional in case of tcmalloc is disabled especially in linux environment because we have already ran the perf test and getting regression.

mohit84 commented:
> > > * to mandate tcmalloc, we have some infra changes to take care of.
> > > * i personally like the support to have different malloc libraries as option in [build, libglusterfs: add --enable-malloc option #2696](https://github.com/gluster/glusterfs/pull/2696)
> > > 
> > > How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.
> > 
> > 
> > If we do this, your first point is still a problem (the infra needs to be updated for sure).
> > For the second point, I would also like to provide support for different malloc libraries specially for debugging and experimenting but, as you said, the number of people that could make use of that (in a useful way) is quite limited, and they are perfectly able to do those changes by themselves (as reverting the mempools patch if they want to). So, why providing an option that we know it will regress if we don't keep mempools around ?
> > @msaju @mohit84 what do you think ?
> 
> I agree with you, we should provide an system_malloc option only while we have mempools. All of us wants to support for different malloc libraries the only thing we need to make mempool optional in case of tcmalloc is disabled especially in linux environment because we have already ran the perf test and getting regression.



> > > * to mandate tcmalloc, we have some infra changes to take care of.
> > > * i personally like the support to have different malloc libraries as option in [build, libglusterfs: add --enable-malloc option #2696](https://github.com/gluster/glusterfs/pull/2696)
> > > 
> > > How about going with your suggestion of failing the configure if tcmalloc dependencies not met (by default, like how we fail for acl, or xattr support). But provide an option to override it if people choose to go with regression. That way, we wouldn't get to 'by mistake' regression.
> > 
> > 
> > If we do this, your first point is still a problem (the infra needs to be updated for sure).
> > For the second point, I would also like to provide support for different malloc libraries specially for debugging and experimenting but, as you said, the number of people that could make use of that (in a useful way) is quite limited, and they are perfectly able to do those changes by themselves (as reverting the mempools patch if they want to). So, why providing an option that we know it will regress if we don't keep mempools around ?
> > @msaju @mohit84 what do you think ?
> 
> I agree with you, we should provide an system_malloc option only while we have mempools. All of us wants to support for different malloc libraries the only thing we need to make mempool optional in case of tcmalloc is disabled especially in linux environment because we have already ran the perf test and getting regression.

Now we can follow approach like below in sequencing manner
1) @dmantipov  please revert the patch #2675
2) Install dependencies for tcmalloc(#https://github.com/gluster/project-infrastructure/issues/140) on all the machines
3) On this patch(#2773) I will disable mempool and enable tcmalloc as a default option
4) @dmantipov will refresh a patch #2696 to provide multiple malloc options and enable mempool in case of system malloc  
 
@xhernandez, @amarts @dmantipov please share if you have any concern

mohit84 committed:
configure: Enable tcmalloc and disable mempool as a default option

The data is already available on github issue(#2771)
specific to getting improvement after enable tcmalloc and
disable mempool so enable tcmalloc and disable mempool
as a default option.

Fixes: #2771
Change-Id: I29256db870185d521d9b9adc2b7fd66646ed1c56
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

deepshikhaaa commented:
/recheck smoke

xhernandez commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

rakshithakamath94 commented:
/recheck smoke

xhernandez commented:
@amarts, are you ok with this change ?

amarts commented:
I would still like to see #2696 get in :+1:

mohit84 commented:
> I would still like to see #2696 get in

Yes we can consider @2696 on top of this.

mohit84 commented:
/run regression

mohit84 commented:
> > I would still like to see #2696 get in
> 
> Yes we can consider @2696 on top of this.

@amarts @xhernandez Please give your vote again, By mistake i have send another patch on the same branch.

mohit84 committed:
iobuf: Call standard allocation api in iobuf if page_size is less than equal 128KB

    During smallfile testing we have observed the performance
    has been improved significantly in case while iobuf use standard
    allocation api to allocate buffer while requested page_size is less
    than or equal to 128KB.The performance data is available at the
    github issue https://github.com/gluster/glusterfs/issues/2771

    Fixes: #2771
    Change-Id: Ib51e54e10c6d7334962e41c8f277a78b676fbe5a
    Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: I4c9ec1d2f4b8df626808fe81b7392314dc8a0e05

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/recheck smoke

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t

2 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
https://build.gluster.org/job/gh_centos7-regression/1604/

mohit84 commented:
Push a patch from another pull request (https://github.com/gluster/glusterfs/pull/2787) so close this pull request
